Make image info available through `ResourceInfo`

bjoeris opened the pull request:
## Description

The image info (layerCount, levelCount, sampleCount, extent, format, type)  was previously only available through `ImageLayouts`. This info is now also stored in `ResourceInfo`.

Storing this in `ResourceInfo` makes this info easily accessible from `VkResourceRecord`s representing `VkImage`s and `VkImageView`s. This simplifies upcoming changes to the image tracking and avoids the lock protecting the `m_ImageLayouts` map.

To simplify this duplication, a new struct `ImageInfo` is added here to hold the image info in both `ResourceInfo` and `ImageLayouts`.

This PR also updates `ImgRefs` to use the new `ImageInfo` class.

bjoeris committed:
Vulkan: Make image info available through `ResourceInfo`

The image info (layerCount, levelCount, sampleCount, extent, format,
type)  was previously only available through `ImageLayouts`. This info
is now also stored in `ResourceInfo`.

Storing this in `ResourceInfo` makes this info easily accessible from
`VkResourceRecord`s representing `VkImage`s and `VkImageView`s. This
simplifies upcoming changes to the image tracking and avoids the lock
protecting the `m_ImageLayouts` map.

To simplify this duplication, a new struct `ImageInfo` is added here to
hold the image info in both `ResourceInfo` and `ImageLayouts`.

Change-Id: I390cbba4c3ce7b0bd6f004521497c21bf4c1e97f

bjoeris committed:
Vulkan: `ImgRefs` now uses `ImageInfo`.

Also added some additional `ImgRefs` tests for 3d images.

Change-Id: I36805ef5135207e691a1c27e6b4fd0903bb8a4f3

baldurk commented on renderdoc/driver/vulkan/vk_serialise.cpp:
```diff
@@ -3418,11 +3418,11 @@ void DoSerialise(SerialiserType &ser, ImageLayouts &el)
+  ser.Serialise(STRING_LITERAL("levelCount"), el.imageInfo.levelCount);
+  ser.Serialise(STRING_LITERAL("sampleCount"), el.imageInfo.sampleCount);
+  ser.Serialise(STRING_LITERAL("extent"), el.imageInfo.extent);
+  ser.Serialise(STRING_LITERAL("format"), el.imageInfo.format);
```
I think it's better to move this into a sub-structure. As long as the members don't change then this will be backwards compatible as the binary format is defined as a linear sequence of bytes with no padding/alignment. So a sub-struct is serialised identically to having those elements linear in the parent struct.

bjoeris replied:
Yes, I was worried about backward compatibility. I switched this to serialize as a substructure, and verified that this does seem to load old captures just fine.

baldurk commented on renderdoc/driver/vulkan/vk_resources.h:
```diff
@@ -1055,19 +1099,16 @@ struct ImgRefs
-      layerCount = extent.depth;
+      this->imageInfo.layerCount = imageInfo.extent.depth;
+    else
+      this->imageInfo.layerCount = imageInfo.layerCount;
```
Is this else needed? It seems redundant after having copied imageInfo already.

bjoeris replied:
Removed.

baldurk reviewed (changes requested):
I think this looks like a good change - I've made a minor comment about adjusting how the struct is serialised to be more idiomatic.

What about the cases where `resInfo` being non-NULL is used as an indicator that the resource is sparse? They don't seem to be changed in this PR.

bjoeris committed:
`VkResourceRecord::resInfo` presence no longer indicates sparse bindings

This VkResourceRecord::resInfo being non-null was previously used as an
indicator that VkBuffer or VkImage resources had sparse bindings. Now,
`resInfo->IsSparse()` must also be true for the resource to be
considered as having sparse bindings.

Change-Id: Ia8a93ecb3248a682f71fcf6455d0d583ca129679

bjoeris commented:
Thanks for catching the missing `resInfo`/sparse bindings change! This has been added.
